File: final_assignment/src/marker_detector.cpp
#include <ros/ros.h>
#include <aruco/aruco.h>
#include <cv_bridge/cv_bridge.h>
#include <actionlib/client/simple_action_client.h>

#include <sensor_msgs/Image.h>
#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Point.h>
#include <final_assignment/RoomInformation.h>
#include <final_assignment/RoomConnection.h>
#include <final_assignment/RoomDoors.h>
#include <final_assignment/OntologyMap.h>
#include <final_assignment/RobotInspectionRoutineAction.h>

aruco::CameraParameters defaultCameraParams;
std::set<int> detectedMarkers;
std::set<int> validatedMarkers;
final_assignment::OntologyMap ontoMapMsg;

ros::ServiceClient roomInfoSrv;
ros::Subscriber imageSub;
ros::Publisher ontoMapPub;
ros::Publisher velPub;

bool retrieveRoomsInformation();
void imageCallback(const sensor_msgs::Image::ConstPtr& msg);
    
    
int main(int argc, char **argv)
{
    // Initializing the ROS Node
    ros::init(argc, argv, "marker_detector");
    ros::NodeHandle n;

    // Initializing variables
    defaultCameraParams = aruco::CameraParameters();
    
    // Creating a Subscriber for the image topic
    imageSub = n.subscribe("/camera/image_raw", 100, imageCallback);
    // Creating a Publisher for the robot velocity
    velPub = n.advertise<geometry_msgs::Twist>("/cmd_vel", 1000, true);
    // Creating a Publisher for the ontology map
    ontoMapPub = n.advertise<final_assignment::OntologyMap>("/ontology_map/build_map", 100, true);
    // Creating a ServiceClient for the rooms information
    roomInfoSrv = n.serviceClient<final_assignment::RoomInformation>("/room_info");
    
    // Creatin a ActionClient for moving the robot arm aroung
    // True causes the client to spin its own thread
    actionlib::SimpleActionClient<final_assignment::RobotInspectionRoutineAction> armAcl("robot_inspection_routine", true);
    
    ROS_INFO("[MARKER DETECTOR] Waiting for InspectionRoutine Server...");
    armAcl.waitForServer();
    ROS_INFO("[MARKER DETECTOR] InspectionRoutine Server found!");
    ROS_INFO("[MARKER DETECTOR] Waiting for MarkerServer Service...");
    roomInfoSrv.waitForExistence();
    ROS_INFO("[MARKER DETECTOR] MarkerServer Service found!");
    
    // Setting the robot velocity very small in order to keep it still
    geometry_msgs::Twist velMsg;
    velMsg.linear.x = 0.001;
    velPub.publish(velMsg);
     
    // Finding all the markers in the room
    do
    {
        // Creating the goal for the movement of the arm
        final_assignment::RobotInspectionRoutineGoal goal;
        armAcl.sendGoal(goal);
    
        // Waiting for the arm movement to complete
        // In the meanwhile, handle callbacks
        ros::Rate rate(10); // 10 hz
        while(!armAcl.getState().isDone())
        {
            ros::spinOnce();
            rate.sleep();
        }
        
        // If the InspectionRoutine was not executed successfully,
        // wait for a bit and then retry
        if(!armAcl.getResult()->completed)
        {
            ROS_INFO("[MARKER DETECTOR] Inspection routine was not executed!");
            ROS_INFO("[MARKER DETECTOR] Waiting for a bit and then retrying!");
            ros::Duration(10).sleep();
            continue;
        }
        
        // Retrieving all the rooms information
        retrieveRoomsInformation();
        ROS_INFO("[MARKER DETECTOR] Validated '%li' markers!", validatedMarkers.size());
        
    }
    while(validatedMarkers.size() < 7);
    
    // Stopping the now unnecessary subscriber for the camera image
    imageSub.shutdown();
    
    // Publishin the ontology map
    ontoMapPub.publish(ontoMapMsg);
    
    // Spinning to prevent stopping the execution
    ros::spin();

    return 0;
}


void imageCallback(const sensor_msgs::Image::ConstPtr& msg)
{
    // Intializing some variables
    aruco::MarkerDetector detector;
    std::vector<aruco::Marker> markers;
    cv_bridge::CvImagePtr cvPtr;
    
    // Creating a CV copy of the image from the camera
    cvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    // Detecting ARUCO markers in the image
    detector.detect(cvPtr->image, markers, defaultCameraParams, 0.05, false);
    
    // Check if a new marker id was detected
    std::pair<std::set<int>::iterator,bool> ret;
    for (std::size_t i = 0; i < markers.size(); ++i)
    {        
        int id = markers.at(i).id;
        // Check if the marker was already validated
        if(validatedMarkers.find(id) != validatedMarkers.end())
            continue;
        // Check if marker was already detected
        if(detectedMarkers.find(id) != detectedMarkers.end())
            continue;
        
        // A new marker has been found
        detectedMarkers.insert(id);
        ROS_INFO("[MARKER DETECTOR] New marker with id '%i' detected!", id);
    }
}


bool retrieveRoomsInformation()
{
    final_assignment::RoomInformation info;
    for(int markerID : detectedMarkers)
    {
        ROS_INFO("[MARKER DETECTOR] Requesting info for ID %i", markerID);
        // Requesting information to the room info service
        info.request.id = markerID;
        if(!roomInfoSrv.call(info))
            return false;
            
        // Check if the detected marker is valid
        if(info.response.room == "no room associated with this marker id")
        {
            ROS_INFO("[MARKER DETECTOR] Detected invalid ID %i, ignoring", markerID);
            continue;
        }
        
        // Storing the room name
        ontoMapMsg.rooms.push_back(info.response.room);
        // Storing the room positions
        geometry_msgs::Point point;
        point.x = info.response.x;
        point.y = info.response.y;
        ontoMapMsg.positions.push_back(point);
        // Storing the room connections
        final_assignment::RoomDoors roomDoors;
        for(final_assignment::RoomConnection conn : info.response.connections)
            roomDoors.doors.push_back(conn.through_door);
        ontoMapMsg.doors.push_back(roomDoors);
        
        // The marker has been parsed as valid
        validatedMarkers.insert(markerID);
    }
    
    // The detected markers have been handled
    detectedMarkers.clear();
    
    return true;
}

File: final_assignment/scripts/marker_detector.py
#!/usr/bin/env python

import rospy
import cv2
import cv2.aruco as aruco
import actionlib

from actionlib import SimpleGoalState
from cv_bridge import CvBridge

from sensor_msgs.msg import Image
from geometry_msgs.msg import Twist
from geometry_msgs.msg import Point
from final_assignment.srv import RoomInformation
from final_assignment.msg import RoomConnection
from final_assignment.msg import RoomDoors
from final_assignment.msg import OntologyMap
from final_assignment.msg import RobotInspectionRoutineAction, RobotInspectionRoutineGoal


class MarkerDetector():

    def __init__(self):
        # Initializing ROS Node
        rospy.init_node("marker_detector")
        # Initializing variables
        self._aruco_params =  aruco.DetectorParameters_create()
        self._cv_bridge = CvBridge()
        self._aruco_dict = aruco.Dictionary_get(aruco.DICT_ARUCO_ORIGINAL)
        self._detected_markers = []
        self._validated_markers = []
        self._onto_map_msg = OntologyMap()
        self._markers_count = rospy.get_param("~markers_count", 7)
        # Creating a Subscriber for the image topic
        self._image_sub = rospy.Subscriber(
            "/camera/image_raw", Image, queue_size=100, 
            callback = self._image_callback
        )
        # Creating a Publisher for the robot velocity
        self._vel_pub = rospy.Publisher(
            "/cmd_vel", Twist, queue_size=1000, latch=True
        )
        # Creating a Publisher for the ontology map
        self._onto_map_pub = rospy.Publisher(
            "/ontology_map/build_map", OntologyMap, queue_size=100, latch=True
        )
        # Creating a ServiceClient for the rooms information
        self._room_info_srv = rospy.ServiceProxy(
            "/room_info", RoomInformation
        )
        # Creating an ActionClient for moving the robot arm
        self._arm_acl = actionlib.SimpleActionClient(
            "/robot_inspection_routine", RobotInspectionRoutineAction
        )
        self._look_for_aruco_markers()


    def _look_for_aruco_markers(self):
        # Waiting for the server to be available
        rospy.loginfo("[MARKER DETECTOR] Waiting for InspectionRoutine Server...")
        self._arm_acl.wait_for_server()
        rospy.loginfo("[MARKER DETECTOR] InspectionRoutine Server found!")
        # Setting the robot velocity very small in order to keep it still
        vel_msg = Twist()
        vel_msg.linear.x = 0.001
        self._vel_pub.publish(vel_msg)
        # Finding all the markers in the room
        while len(self._validated_markers) < self._markers_count :
            # Creating a goal for the movement of the arm
            inspection_goal = RobotInspectionRoutineGoal()
            self._arm_acl.send_goal(inspection_goal)
            # Waiting for the arm movement to complete
            self._arm_acl.wait_for_result()
            # Check if the arm movement was complete
            if not self._arm_acl.get_result().completed :
                rospy.logwarn("[MARKER DETECTOR] Inspection routine was not executed!")
                rospy.logwarn("[MARKER DETECTOR] Waiting for a bit and then retrying!")
                rospy.sleep(10)
                continue
            # Retrieving all the rooms information
            self._retrieve_rooms_information()
            rospy.loginfo("[MARKER DETECTOR] Validated "+str(len(self._validated_markers))+" markers!")
        # Stopping the subscriber for the camera image
        self._image_sub.unregister()
        # Publishing the ontology map
        self._onto_map_pub.publish(self._onto_map_msg)


    def _image_callback(self, msg):
        # Encoding the image using cv
        image = self._cv_bridge.imgmsg_to_cv2(msg, "bgr8")
        # Getting the markers from aruco
        corners, ids, _ = aruco.detectMarkers(image, self._aruco_dict, parameters=self._aruco_params)
        if len(corners) == 0 :
            return
        # Adding all the new detected ids to the set
        for _id in ids.flatten() :
            # Check if the marker id was already validated
            if _id in self._validated_markers :
                continue
            # Check if the marker was already detected
            if _id in self._detected_markers :
                continue
            # A new marker has been found
            self._detected_markers.append(_id)
            rospy.loginfo("[MARKER DETECTOR] New marker with id "+str(_id)+" detected!")
    

    def _retrieve_rooms_information(self):
        # Waiting for the service to be available
        rospy.loginfo("[MARKER DETECTOR] Waiting for MarkerServer Service...")
        self._room_info_srv.wait_for_service()
        rospy.loginfo("[MARKER DETECTOR] MarkerServer Service found!")
        # Creating a new room information message
        room_info_msg = RoomInformation()
        for _id in self._detected_markers :
            # Requesting information to the room service
            rospy.loginfo("[MARKER DETECTOR] Requesting info for ID "+str(_id))
            response = self._room_info_srv(_id)
            # Check if the id was valid
            if response.room == "no room associated with this marker id":
                rospy.loginfo("[MARKER DETECTOR] Detected marker with invalid ID "+str(_id)+", ignoring")
                continue
            # Storing the room name
            self._onto_map_msg.rooms.append(response.room)
            # Storing the room positions
            point = Point()
            point.x = response.x
            point.y = response.y
            self._onto_map_msg.positions.append(point)
            # Storing the room connections
            room_doors = RoomDoors()
            for connection in response.connections :
                room_doors.doors.append(connection.through_door)
            self._onto_map_msg.doors.append(room_doors)
            # The marker has been parsed as valid
            self._validated_markers.append(_id)
        # The detected markers have been handled
        self._detected_markers.clear()


if __name__ == "__main__":
    MarkerDetector()
    rospy.spin()

